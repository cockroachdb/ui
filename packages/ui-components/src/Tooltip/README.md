[back to components](../README.md)

# Tooltip

A Tooltip is a styled popover element that can be shown on hover.

```javascript
import React from "react";
import { Tooltip } from "@cockroachlabs/ui-components";

export const SomeExampleReact = () => (
  <Tooltip content={"Tooltip content"}>
    <div id="hoverme">hover me</div>
  </Tooltip>
);
```

## Properties

These are the properties that can be provided to the `Tooltip` component.

### `children: ReactNode | text`

The content hover on what will trigger display of tooltip.

This prop is **required**.

### `position?: Enum("left", "left-start", "left-end", "right", "right-start", "right-end", "top", "top-start", "top-end", "bottom", "bottom-start","bottom-end")`

**default value: `bottom`**

Placement of tooltip around anchor content.

This prop is **optional**.

### `style?: Enum("default", "light", "dark", "tableTitle")`

**default value: `default`**

Will alter the visual style of the Tooltip by changing the `background-color`, `color`, `border-color` properties of the rendered element.

This prop is **optional**.

### `visible?: boolean`

**default value: `false`**

Force tooltip stay visible.

This prop is **optional**.