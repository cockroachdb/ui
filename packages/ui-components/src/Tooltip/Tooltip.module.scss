@import "../styles/tokens.scss";

//
// Tooltips
// --------------------------------------------------

$border-radius-base: 3px;
$tooltip-border-width: 1px;
$tooltip-shadow-width: 0px;

$tooltip-arrow-width: 10px;
$tooltip-arrow-half-width: 5px;

// Base class
.Tooltip {
  all: initial;
  visibility: hidden;
  opacity: 0;
  position: absolute;
  z-index: -1;
  transition: opacity 0.1s, visibility 0.1s, z-index 0.1s;
  min-width: 300px;
  max-width: 500px;

  p,
  pre,
  code {
    all: initial;
    font-family: $type-family-ui;
    font-size: $type-size-font-medium;
  }

  & > .content {
    @include crl-depth(1);
    padding: 8px 10px;
    margin: 30px;
    text-align: left;
    text-decoration: none;
    border-radius: $border-radius-base;
    border: $tooltip-border-width solid;
    position: relative;
    overflow: visible;
    font-family: $type-family-ui;
  }

  &[data-show] {
    visibility: visible;
    opacity: 1;
    z-index: 9999;
    transition: opacity 0.1s, visibility 0.1s, z-index 0.1s;
  }

  .arrow,
  .arrowInner {
    position: absolute;
    width: $tooltip-arrow-width;
    height: $tooltip-arrow-width;
    overflow: visible;
    &:before {
      content: "";
      position: absolute;
      width: $tooltip-arrow-width;
      height: $tooltip-arrow-width;
      transform: rotate(45deg);
      border: $tooltip-border-width solid;
    }
  }

  .arrow:before {
    @include crl-depth(1);
    z-index: -5;
  }

  .arrowInner:before {
    z-index: 5;
  }

  &[data-popper-placement^="top"] .arrow,
  &[data-popper-placement^="top"] .arrowInner {
    bottom: -$tooltip-arrow-half-width;
  }

  &[data-popper-placement^="bottom"] .arrow,
  &[data-popper-placement^="bottom"] .arrowInner {
    top: -$tooltip-arrow-half-width;
  }

  &[data-popper-placement^="left"] .arrow,
  &[data-popper-placement^="left"] .arrowInner {
    right: -$tooltip-arrow-half-width;
  }

  &[data-popper-placement^="right"] .arrow,
  &[data-popper-placement^="right"] .arrowInner {
    left: -$tooltip-arrow-half-width;
  }
}

@mixin crl-tooltip-style($bg, $border, $color) {
  .content {
    border-color: $border;
    background-color: $bg;
    color: $color;
    p,
    pre,
    code {
      color: $color;
    }
  }
  .arrow:before {
    border-color: $border;
    background-color: $bg;
  }
  .arrowInner:before {
    background-color: $bg;
    border-color: transparent;
  }
}

//
// Default style
// --------------------------------------------------
.default {
  @include crl-tooltip-style(
    $color-core-neutral-6,
    $color-core-neutral-6,
    $color-base-white
  );
}
//
// Dark style
// --------------------------------------------------
.dark {
  @include crl-tooltip-style(
    $color-base-black,
    $color-base-black,
    $color-base-white
  );
}

//
// TableTitle
// --------------------------------------------------
.tableTitle {
  @include crl-tooltip-style(
    $color-core-neutral-6,
    $color-core-neutral-6,
    $color-base-white
  );
  .text {
    &,
    p, a {
      font-weight: 600;
      font-size: $type-size-font-small;
      line-height: $type-size-line-height-small;
      letter-spacing: 0.3px;
      color: $color-intent-neutral-0;
    }

    a {
      text-decoration: underline;
      &:hover {
        opacity: 0.7;
        color: $color-intent-neutral-0;
      }
    }
    code {
      font-family: $type-family-code;
      font-size: $type-size-font-small;
      line-height: $type-size-line-height-regular;
      color: $color-intent-neutral-8;
      background: $color-intent-neutral-0;
      border-radius: 3px;
      padding: 2px 5px;
    }
  }
}

//
// Light style
// --------------------------------------------------
.light {
  @include crl-tooltip-style(
    $color-base-white,
    $color-core-neutral-2,
    $color-font-3
  );
}
