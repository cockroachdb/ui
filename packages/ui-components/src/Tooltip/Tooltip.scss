@import "../styles/tokens.scss";

//
// Tooltips
// --------------------------------------------------

$border-radius-base: 3px;
$tooltip-border-width: 1px;
$tooltip-shadow-width: 0px;
$tooltip-arrow-width: 6px;

@mixin crl-tooltip-style($bg, $border, $color) {
  .crl-tooltip-inner {
    border-color: $border;
    background-color: $bg;
    color: $color;
  }
  .crl-tooltip-arrow {
    border-color: $border;
  }
  .crl-tooltip-arrow-inner {
    border-color: $bg;
  }
}

// Base class
.crl-tooltip {
  //
  // Default style
  // --------------------------------------------------
  &.default {
    @include crl-tooltip-style($crl-neutral-6, $crl-neutral-6, $crl-base-white);
  }
  //
  // Dark style
  // --------------------------------------------------
  &.dark {
    @include crl-tooltip-style(
      $crl-base-black,
      $crl-base-black,
      $crl-base-white
    );
  }
  //
  // Light style
  // --------------------------------------------------
  &.light {
    @include crl-tooltip-style($crl-base-white, $crl-neutral-2, $crl-base-text);
  }

  position: absolute;
  z-index: 1070;
  display: block;
  visibility: visible;
  padding: $tooltip-shadow-width;
  &-hidden {
    display: none;
  }
  // Wrapper for the tooltip content
  &-inner {
    @include crl-depth(1);
    padding: 8px 10px;
    text-align: left;
    text-decoration: none;
    border-radius: $border-radius-base;
    min-height: 24px;
    max-width: 300px;
    border: $tooltip-border-width solid;
  }

  // Arrows
  &-arrow,
  &-arrow-inner {
    position: absolute;
    width: 0;
    height: 0;
    border-style: solid;
  }

  &-placement-top &-arrow,
  &-placement-topLeft &-arrow,
  &-placement-topRight &-arrow {
    bottom: -$tooltip-arrow-width + $tooltip-shadow-width;
    margin-left: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width $tooltip-arrow-width 0;
    border-left-color: transparent !important;
    border-right-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-top &-arrow-inner,
  &-placement-topLeft &-arrow-inner,
  &-placement-topRight &-arrow-inner {
    bottom: $tooltip-border-width;
    margin-left: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width $tooltip-arrow-width 0;
    border-left-color: transparent !important;
    border-right-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-top &-arrow {
    left: 50%;
  }

  &-placement-topLeft &-arrow {
    left: 15%;
  }

  &-placement-topRight &-arrow {
    right: 15%;
  }

  &-placement-right &-arrow,
  &-placement-rightTop &-arrow,
  &-placement-rightBottom &-arrow {
    left: -$tooltip-arrow-width + $tooltip-shadow-width;
    margin-top: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width $tooltip-arrow-width $tooltip-arrow-width
      0;
    border-left-color: transparent !important;
    border-top-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-right &-arrow-inner,
  &-placement-rightTop &-arrow-inner,
  &-placement-rightBottom &-arrow-inner {
    left: $tooltip-border-width;
    margin-top: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width $tooltip-arrow-width $tooltip-arrow-width
      0;
    border-left-color: transparent !important;
    border-top-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-right &-arrow {
    top: 50%;
  }

  &-placement-rightTop &-arrow {
    top: 15%;
    margin-top: 0;
  }

  &-placement-rightBottom &-arrow {
    bottom: 15%;
  }

  &-placement-left &-arrow,
  &-placement-leftTop &-arrow,
  &-placement-leftBottom &-arrow {
    right: -$tooltip-arrow-width + $tooltip-shadow-width;
    margin-top: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width 0 $tooltip-arrow-width
      $tooltip-arrow-width;
    border-right-color: transparent !important;
    border-top-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-left &-arrow-inner,
  &-placement-leftTop &-arrow-inner,
  &-placement-leftBottom &-arrow-inner {
    right: $tooltip-border-width;
    margin-top: -$tooltip-arrow-width;
    border-width: $tooltip-arrow-width 0 $tooltip-arrow-width
      $tooltip-arrow-width;
    border-right-color: transparent !important;
    border-top-color: transparent !important;
    border-bottom-color: transparent !important;
  }

  &-placement-left &-arrow {
    top: 50%;
  }

  &-placement-leftTop &-arrow {
    top: 15%;
    margin-top: 0;
  }

  &-placement-leftBottom &-arrow {
    bottom: 15%;
  }

  &-placement-bottom &-arrow,
  &-placement-bottomLeft &-arrow,
  &-placement-bottomRight &-arrow {
    top: -$tooltip-arrow-width + $tooltip-shadow-width;
    margin-left: -$tooltip-arrow-width;
    border-width: 0 $tooltip-arrow-width $tooltip-arrow-width;
    border-right-color: transparent !important;
    border-top-color: transparent !important;
    border-left-color: transparent !important;
  }

  &-placement-bottom &-arrow-inner,
  &-placement-bottomLeft &-arrow-inner,
  &-placement-bottomRight &-arrow-inner {
    top: $tooltip-border-width;
    margin-left: -$tooltip-arrow-width;
    border-width: 0 $tooltip-arrow-width $tooltip-arrow-width;
    border-right-color: transparent !important;
    border-top-color: transparent !important;
    border-left-color: transparent !important;
  }

  &-placement-bottom &-arrow {
    left: 50%;
  }

  &-placement-bottomLeft &-arrow {
    left: 15%;
  }

  &-placement-bottomRight &-arrow {
    right: 15%;
  }
}
