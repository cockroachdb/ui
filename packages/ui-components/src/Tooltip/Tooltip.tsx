import React, {
  FunctionComponent,
  ReactElement,
  useEffect,
  useState,
} from "react";
import { usePopper } from "react-popper";
import classNames from "classnames/bind";
import css from "./Tooltip.module.scss";

export type TooltipPosition =
  | "left"
  | "left-start"
  | "left-end"
  | "right"
  | "right-start"
  | "right-end"
  | "top"
  | "top-start"
  | "top-end"
  | "bottom"
  | "bottom-start"
  | "bottom-end";

export type TooltipStyle = "default" | "light" | "dark" | "tableTitle";

export interface TooltipProps {
  placement?: TooltipPosition;
  style?: TooltipStyle;
  content: ReactElement | string;
  visible?: boolean;
  timeout?: number;
}

export const Tooltip: FunctionComponent<TooltipProps> = ({
  placement = "bottom",
  style = "default",
  children,
  content,
  visible = false,
  timeout = 500,
}) => {
  const [referenceElement, setReferenceElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);
  const [arrowElement, setArrowElement] = useState(null);

  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    placement,
    modifiers: [
      { name: "arrow", options: { element: arrowElement } },
      { name: "offset", options: { offset: [0, -25] } },
      {
        name: "computeStyles",
        options: {
          gpuAcceleration: false, // true by default
        },
      },
    ],
  });

  useEffect(() => {
    const onScroll = () => {
      popperElement.removeAttribute("data-show");
    };
    window.addEventListener("scroll", onScroll);
    return () => window.removeEventListener("scroll", onScroll);
  }, [popperElement]);

  if (!content) {
    return <>{children}</>;
  }

  const classnames = classNames(css[style], css.Tooltip);
  if (visible && popperElement) {
    popperElement.setAttribute("data-show", "");
  }

  const wrappedChildren = React.Children.map(children, child => {
    return React.cloneElement(
      typeof children === "string" ? <>{child}</> : (child as ReactElement),
    );
  });

  const timers: Array<ReturnType<typeof setTimeout>> = [];
  return (
    <div
      ref={setReferenceElement}
      onMouseOver={() => {
        const t = setTimeout(() => {
          popperElement.setAttribute("data-show", "");
        }, timeout);
        timers.push(t);
      }}
      onMouseLeave={() => {
        timers.forEach((t: ReturnType<typeof setTimeout>) => {
          clearTimeout(t);
          popperElement.removeAttribute("data-show");
        });
      }}
    >
      {wrappedChildren}

      <div
        className={classnames}
        ref={setPopperElement}
        style={styles.popper}
        data-jest="tooltip"
        onMouseLeave={() => {
          if (visible) return;
          popperElement.removeAttribute("data-show");
        }}
        onClick={e => {
          e.stopPropagation();
        }}
        {...attributes.popper}
      >
        <div className={css.content}>
          <div
            className={css.arrow}
            ref={setArrowElement}
            style={styles.arrow}
          />
          <div className={css.arrowInner} style={styles.arrow} />
          <div className={css.text}>{content}</div>
        </div>
      </div>
    </div>
  );
};
