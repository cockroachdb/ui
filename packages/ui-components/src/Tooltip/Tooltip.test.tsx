import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";

import { Tooltip } from "./Tooltip";

const TRIGGER_TEXT = "hover trigger";
const HOVER_TEXT = "hover text";

const TooltipExample = (
  <Tooltip content={<div id="content">{HOVER_TEXT}</div>}>
    <div>{TRIGGER_TEXT}</div>
  </Tooltip>
);

describe("Tooltip", () => {
  it("is shown on hover", () => {
    render(TooltipExample);
    expect(screen.getByText(HOVER_TEXT)).not.toHaveAttribute("data-show");
    fireEvent.mouseOver(screen.getByText(TRIGGER_TEXT));
    waitFor(() => {
      expect(screen.getByText(HOVER_TEXT)).toHaveAttribute("data-show");
    });
  });

  it("is not shown on hover if content is empty", () => {
    const { container } = render(
      <Tooltip content={undefined}>
        <div>{TRIGGER_TEXT}</div>
      </Tooltip>,
    );
    expect(container.querySelector('[data-jest="tooltip"]')).toBeNull();
    fireEvent.mouseOver(screen.getByText(TRIGGER_TEXT));
    expect(container.querySelector('[data-jest="tooltip"]')).toBeNull();
  });
});
