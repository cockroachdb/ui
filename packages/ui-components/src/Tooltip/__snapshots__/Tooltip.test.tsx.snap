// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
<div
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    id="hoverme"
    key=".0"
  >
    hover me
  </div>
  <div
    className="function classNames () {
		var classes = [];

		for (var i = 0; i < arguments.length; i++) {
			var arg = arguments[i];
			if (!arg) continue;

			var argType = typeof arg;

			if (argType === 'string' || argType === 'number') {
				classes.push(this && this[arg] || arg);
			} else if (Array.isArray(arg)) {
				classes.push(classNames.apply(this, arg));
			} else if (argType === 'object') {
				if (arg.toString === Object.prototype.toString) {
					for (var key in arg) {
						if (hasOwn.call(arg, key) && arg[key]) {
							classes.push(this && this[key] || key);
						}
					}
				} else {
					classes.push(arg.toString());
				}
			}
		}

		return classes.join(' ');
	} Tooltip"
    data-jest="tooltip"
    onClick={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
  >
    <div
      className="content"
    >
      <div
        className="arrow"
        style={
          Object {
            "position": "absolute",
          }
        }
      />
      <div
        className="arrowInner"
        style={
          Object {
            "position": "absolute",
          }
        }
      />
      <div
        className="text"
      >
        <div
          id="content"
        >
          test
        </div>
      </div>
    </div>
  </div>
</div>
`;
